feat: end-2-end mqttt bridge health check

pudkrong opened the pull request:
It will use simulator device to publish message to nRF Cloud and wait for incoming message from web socket for 10s.

pudkrong committed:
feat: end-2-end mqttt bridge health check

coderbyheart commented on cdk/backend.ts:
```diff
@@ -40,6 +40,8 @@ const packagesInLayer: string[] = [
 	'lodash-es',
 	'@middy/core',
+	'mqtt',
+	'ws',
```
Put in separate layer, these packages are quite large and not needed in the other lambdas.

coderbyheart commented on cdk/backend.ts:
```diff
@@ -83,6 +86,7 @@ new BackendApp({
 	iotEndpoint: await getIoTEndpoint({ iot })(),
 	mqttBridgeCertificate,
 	caCertificate,
+	amazonRootCA1,
```
No need to pass files in source code as parameters.

coderbyheart commented on lambda/healthCheck.ts:
```diff
@@ -0,0 +1,190 @@
+const { DevicesTableName, stackName, amazonRootCA1, websocketUrl } = fromEnv({
+	DevicesTableName: 'DEVICES_TABLE_NAME',
+	stackName: 'STACK_NAME',
+	amazonRootCA1: 'AMAZON_ROOT_CA1',
```
Include as string in source.

coderbyheart commented on cdk/resources/Integration.ts:
```diff
@@ -277,5 +298,45 @@ export class Integration extends Construct {
 			'inbound-mqtt',
 		)
+
+		const scheduler = new Events.Rule(this, 'scheduler', {
```
Move to separate construct.

coderbyheart commented on lambda/healthCheck.ts:
```diff
@@ -0,0 +1,190 @@
+	async (deviceId: string, message: Record<string, unknown>) => {
+		await new Promise((resolve, reject) => {
+			const mqttClient = mqtt.connect({
+				host: bridgeInfo.mqttEndpoint,
```
Device should publish message to nRF Cloud MQTT API. Because otherwise we are not testing that the bridge has correctly subscribed to the nRF MQTT API.

coderbyheart commented on lambda/healthCheck.ts:
```diff
@@ -0,0 +1,190 @@
+	reject: (reason: any) => void
+}
+
+const defer = (timeout: number): ReturnDefer<any> => {
+	const ret = {} as ReturnDefer<any>
+	const timer = setTimeout(() => {
+		ret.reject('timeout')
+	}, timeout)
+
+	const promise = new Promise<any>((_resolve, _reject) => {
+		ret.resolve = (v) => {
+			clearTimeout(timer)
+			_resolve(v)
+		}
+		ret.reject = (reason) => {
+			clearTimeout(timer)
+			_reject(reason)
+		}
+	})
+
+	ret.promise = promise
+
+	return ret
+}
```
Utility functions needs tests.

coderbyheart commented on lambda/healthCheck.ts:
```diff
@@ -0,0 +1,190 @@
+				}
+			},
+		})
+		metrics.addMetric('success', MetricUnits.Count, 1)
```
Record time difference in seconds between `ts` and now.

pudkrong committed:
refactor: move health check into its own stack

coderbyheart commented on cdk/resources/HealthCheckMqttBridge.ts:
```diff
@@ -0,0 +1,97 @@
+				STACK_NAME: Stack.of(this).stackName,
+				DEVICES_TABLE_NAME: deviceStorage.devicesTable.tableName,
+				WEBSOCKET_URL: websocketAPI.websocketURI,
+				AMAZON_ROOT_CA1:
+					'-----BEGIN CERTIFICATE-----\n' +
+					'MIIDQTCCAimgAwIBAgITBmyfz5m/jAo54vB4ikPmljZbyjANBgkqhkiG9w0BAQsF\n' +
+					'ADA5MQswCQYDVQQGEwJVUzEPMA0GA1UEChMGQW1hem9uMRkwFwYDVQQDExBBbWF6\n' +
+					'b24gUm9vdCBDQSAxMB4XDTE1MDUyNjAwMDAwMFoXDTM4MDExNzAwMDAwMFowOTEL\n' +
+					'MAkGA1UEBhMCVVMxDzANBgNVBAoTBkFtYXpvbjEZMBcGA1UEAxMQQW1hem9uIFJv\n' +
+					'b3QgQ0EgMTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBALJ4gHHKeNXj\n' +
+					'ca9HgFB0fW7Y14h29Jlo91ghYPl0hAEvrAIthtOgQ3pOsqTQNroBvo3bSMgHFzZM\n' +
+					'9O6II8c+6zf1tRn4SWiw3te5djgdYZ6k/oI2peVKVuRF4fn9tBb6dNqcmzU5L/qw\n' +
+					'IFAGbHrQgLKm+a/sRxmPUDgH3KKHOVj4utWp+UhnMJbulHheb4mjUcAwhmahRWa6\n' +
+					'VOujw5H5SNz/0egwLX0tdHA114gk957EWW67c4cX8jJGKLhD+rcdqsq08p8kDi1L\n' +
+					'93FcXmn/6pUCyziKrlA4b9v7LWIbxcceVOF34GfID5yHI9Y/QCB/IIDEgEw+OyQm\n' +
+					'jgSubJrIqg0CAwEAAaNCMEAwDwYDVR0TAQH/BAUwAwEB/zAOBgNVHQ8BAf8EBAMC\n' +
+					'AYYwHQYDVR0OBBYEFIQYzIU07LwMlJQuCFmcx7IQTgoIMA0GCSqGSIb3DQEBCwUA\n' +
+					'A4IBAQCY8jdaQZChGsV2USggNiMOruYou6r4lK5IpDB/G/wkjUu0yKGX9rbxenDI\n' +
+					'U5PMCCjjmCXPI6T53iHTfIUJrU6adTrCC2qJeHZERxhlbI1Bjjt/msv0tadQ1wUs\n' +
+					'N+gDS63pYaACbvXy8MWy7Vu33PqUXHeeE6V/Uq2V8viTO96LXFvKWlJbYK8U90vv\n' +
+					'o/ufQJVtMVT8QtPHRh8jrdkPSHCa2XV4cdFyQzR1bldZwgJcJmApzyMZFo6IQ6XU\n' +
+					'5MsI+yMRQ+hDKXJioaldXgjUkK642M4UwtBV8ob2xJNDd2ZhwLnoQdeXeGADbkpy\n' +
+					'rqXRfboQnoZsG4q5WTP468SQvvG5\n' +
+					'-----END CERTIFICATE-----\n',
+			},
```
hardcode in lambda source code

coderbyheart commented on lambda/healthCheck.ts:
```diff
@@ -0,0 +1,177 @@
+			protocolVersion: 4,
+			clean: true,
+			clientId: deviceId,
+			key: nrfCloudInfo.accountDevicePrivateKey,
+			cert: nrfCloudInfo.accountDeviceClientCert,
```
Account devices to not send data, it's individual devices.

coderbyheart replied:
Will also be relevant when we want to have shadow fetcher health check.

pudkrong committed:
refactor: hard code amazon root cert into source code

pudkrong committed:
feat: create fake health check device

pudkrong committed:
fix: delete parameter store

When using DeleteParametersCommand to delete parameters more than 10,
it will error as Member must have length less than or equal to 10

pudkrong committed:
feat: add create health check device cli

pudkrong committed:
Merge branch 'saga' into 3-message-bridge-stops-working-after-a-while

pudkrong commented on cli/cli.ts:
```diff
@@ -74,6 +82,13 @@ const CLI = async ({ isCI }: { isCI: boolean }) => {
 				stackName: STACK_NAME,
 			}),
 		)
+		commands.push(
```
I have added a new command `create-health-check-device` for end-to-end health check purpose. We need to manually call this command once to enable the feature. @coderbyheart

pudkrong commented on README.md:
````diff
@@ -33,6 +33,7 @@ need to prepare nRF Cloud API key.
 ```bash
 ./cli.sh configure thirdParty nrfcloud apiKey <API key>
 ./cli.sh initialize-nrfcloud-account
+./cli.sh create-health-check-device
````
@coderbyheart health check device is created here
